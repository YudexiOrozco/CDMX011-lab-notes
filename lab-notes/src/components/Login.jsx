import React from 'react'
import logo from '../images/logo.png'
import cook from '../images/cooking.jpg'
import { auth, db } from '../firebase'
import {withRouter} from 'react-router-dom'

const Login = (props) => {

  const [email, setEmail] = React.useState('')
  const [password, setPassword] = React.useState('')
  const [error, setError] = React.useState(null)
  const [registro, setRegistro] = React.useState(false)

  const procesarDatos = (e) =>{
    e.preventDefault()
    if(!email.trim()){
      // console.log('Ingrese Email')
      setError('Ingrese Email')
      return
    }
    if(!password.trim()){
      // console.log('Ingrese Password')
      setError('Ingrese Password')
      return
    }
    if(password.length < 6){
      // console.log('Password mayor a 6 caracteres')
      setError('Password mayor a 6 caracteres')
    }
    setError(null)
    console.log('Pasando todas las validaciones!')

    if(registro){
      registrar()
    }else{
      login()
    }
  }

  const login = React.useCallback(async() => {
    try {
      await auth.signInWithEmailAndPassword(email, password)
      // console.log(res.user)
      setEmail('')
      setPassword('')
      setError(null)
      props.history.push('/wall')
    } catch (error) {
      console.log(error)
      if(error.code === 'auth/user-not-found'){
        setError('Email not registered...')
      }
      if(error.code === 'auth/wrong-password'){
        setError('The password is invalid')
      }
    }

  }, [email, password, props.history])

  const registrar =React.useCallback(async() => {

    try {
      const res = await auth.createUserWithEmailAndPassword(email, password)
      // console.log(res.user)
      await db.collection('usuarios').doc(res.user.email).set({
        email: res.user.email,
        uid: res.user.uid
      })
      setEmail('')
      setPassword('')
      setError(null)
      props.history.push('/wall')
    } catch (error) {
      console.log(error)
      if(error.code === 'auth/invalid-email'){
        setError('The email address is incorrect')
      }
      if(error.code === 'auth/email-already-in-use'){
        setError('The email is already in use')
      }
    }

  }, [email, password, props.history])

  return (
    <div className='container-login'>
      <img className='logo' src={logo} alt='logo app foodpad'/>
      <img className='cook' src={cook} alt='cooking'/>
      <h3 className='slogan'>Manage, organize and save your favorite recipes</h3>
      <h4 className='text-register'>
        {
          registro ? 'Registro de usuarios' : 'Login de acceso'
        }
      </h4>
      <div className='container'>
        <div className='container-form'>
          <form onSubmit={procesarDatos}>
            {
              error && (
                <div>{error}</div>
              )
            }
            <input 
              type = 'email' 
              className = 'inputs'
              placeholder = 'Ingrese un email'
              onChange = {e => setEmail(e.target.value)}
              value = {email}
            />
            <input 
              type = 'password' 
              className = 'inputs'
              placeholder = 'Ingrese una contraseña'
              onChange = {e => setPassword(e.target.value)}
              value = {password}
            />
            <button className='btn-login' type='submit'>
              {
                registro ? 'Registrarse' : 'Acceder'
              }
            </button>
            <button 
              className='btn-login' 
              onClick={() => setRegistro(!registro)}
              type='button'
            >
             {
               registro ? '¿Ya estas registrado?' : '¿No tienes cuenta?'
             }
            </button>
          </form>
        </div>
      </div>
    </div>
  )
}

export default withRouter(Login)
