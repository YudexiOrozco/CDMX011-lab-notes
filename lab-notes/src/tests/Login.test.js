import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { shallow } from "enzyme";
import Login from '../components/Login';

describe('Pruebas en el <Login />', () => {
  let wrapper = shallow(  <Login />  );
 
  beforeEach( () => {
    
    wrapper = shallow(  <Login />  );
    
  });

  test( 'Sobre etiqueta span dentro de label', () => {
    render(<Login />)
    const contentEmail = screen.getByText('Email')
    expect(contentEmail).toBeInTheDocument()
  });

  test('Debe de mostrar el componente <Login /> correctamente', () => {

    expect( wrapper ).toMatchSnapshot();

  })

  test('Mostrar error si se envia el formulario con los inputs vacios', () => {
  
    wrapper.find('form').simulate('submit', { preventDefault(){} });
    const text = wrapper.find('p').text();

    expect( text ).toBe('Ingrese Email');
  })
  
  test('Mostrar error si se envia el formulario con el password vacio', () => {
  
    let input = wrapper.find('input').at(0);
    input.simulate('change', { target: { value: 'test@test.com' }});

    wrapper.find('form').simulate('submit', { preventDefault(){} });
    const text = wrapper.find('p').text();

    expect( text ).toBe('Ingrese Password');
  })

  test('El texto de botones debe cambiar cuando se hace click en New Account', () => {
  
    wrapper.find('button').at(1).simulate('click');

    let inputs = wrapper.find('span');

    expect(inputs.at(0).text() ).toBe('Email');
    expect(inputs.at(1).text() ).toBe('Password');

    expect(wrapper.find('input').length).toBe(2)

    const buttons =  wrapper.find('button')

    expect(buttons.at(0).text() ).toBe('Sign Up');
    expect(buttons.at(1).text() ).toBe('You already sign up?');
  })

  test('El texto de botones debe cambiar cuando se hace click en You already sign up?', () => {
  
    wrapper.find('button').at(1).simulate('click'); // New Account
    wrapper.find('button').at(1).simulate('click'); // You already sign up?'

    let inputs = wrapper.find('span');

    expect(inputs.at(0).text() ).toBe('Email');
    expect(inputs.at(1).text() ).toBe('Password');

    expect(wrapper.find('input').length).toBe(2)

    const buttons =  wrapper.find('button');

    expect(buttons.at(0).text() ).toBe('Log In');
    expect(buttons.at(1).text() ).toBe('New account');
  })
})
