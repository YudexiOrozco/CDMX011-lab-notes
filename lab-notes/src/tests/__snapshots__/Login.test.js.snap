// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <Login /> Debe de mostrar el componente <Login /> correctamente 1`] = `
<div
  className="container-login"
>
  <img
    alt="logo app foodpad"
    className="logo-login"
    src="logo.png"
  />
  <img
    alt="cooking"
    className="cook"
    src="cooking.png"
  />
  <h3
    className="slogan"
  >
    Manage, organize and save your favorite recipes.
  </h3>
  <div
    className="container"
  >
    <div
      className="container-form"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="div-input"
        >
          <label
            className="inputs"
          >
            <span
              className="span-email"
            >
              Email
            </span>
            <input
              className="input-email"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              type="email"
              value=""
            />
          </label>
          <label
            className="inputs"
          >
            <span
              className="span-email"
            >
              Password
            </span>
            <input
              className="input-email"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              type="password"
              value=""
            />
          </label>
        </div>
        <button
          className="btn-login"
          type="submit"
        >
          Log In
        </button>
        <button
          className="btn-new"
          onClick={[Function]}
          type="button"
        >
          New account
        </button>
      </form>
    </div>
  </div>
  <img
    alt="vector"
    className="vector"
    src="vector.png"
  />
</div>
`;
